Mathlib/Data/Sign.lean

```diff
@@ -0,0 +1,523 @@
+113488-general/topic/type.20class.20inference.20issues/near/276937942 -/
+attribute [local instance] LinearOrderedRing.decidable_lt
+
+theorem sign_mul (x y : α) : sign (x * y) = sign x * sign y := by
+  rcases lt_trichotomy x 0 with (hx | hx | hx) <;> rcases lt_trichotomy y 0 with (hy | hy | hy) <;>
+    simp [hx, hy, mul_pos_of_neg_of_neg, mul_neg_of_neg_of_pos, mul_neg_of_pos_of_neg]
+#align sign_mul sign_mul
+
+/-- `sign` as a `MonoidWithZeroHom` for a nontrivial ordered semiring. Note that linearity
+is required; consider ℂ with the order `z ≤ w` iff they have the same imaginary part and
+`z - w ≤ 0` in the reals; then `1 + i` and `1 - i` are incomparable to zero, and thus we have:
+`0 * 0 = sign (1 + i) * sign (1 - i) ≠ sign 2 = 1`. (`Complex.orderedCommRing`) -/
+def signHom : α →*₀ SignType where
+  toFun := sign
+  map_zero' := sign_zero
+  map_one' := sign_one
+  map_mul' := sign_mul
+#align sign_hom signHom
+
+theorem sign_pow (x : α) (n : ℕ) : sign (x ^ n) = sign x ^ n := by
+  change signHom (x ^ n) = signHom x ^ n
+  exact map_pow _ _ _
+#align sign_pow sign_pow
+
+end LinearOrderedRing
+
+section AddGroup
+
+variable [AddGroup α] [Preorder α] [DecidableRel ((· < ·) : α → α → Prop)]
+
+theorem Left.sign_neg [CovariantClass α α (· + ·) (· < ·)] (a : α) : sign (-a) = -sign a := by
+  simp_rw [sign_apply, Left.neg_pos_iff, Left.neg_neg_iff]
+  split_ifs with h h'
+  · exact False.elim (lt_asymm h h')
+  · simp
+  · simp
+  · simp
+#align left.sign_neg Left.sign_neg
+
+theorem Right.sign_neg [CovariantClass α α (Function.swap (· + ·)) (· < ·)] (a : α) :
+    sign (-a) = -sign a := by
+  simp_rw [sign_apply, Right.neg_pos_iff, Right.neg_neg_iff]
+  split_ifs with h h'
+  · exact False.elim (lt_asymm h h')
+  · simp
+  · simp
+  · simp
+#align right.sign_neg Right.sign_neg
+
+end AddGroup
+
+section LinearOrderedAddCommGroup
+open BigOperators
+
+variable [LinearOrderedAddCommGroup α]
+
+/- I'm not sure why this is necessary, see
+https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Decidable.20vs.20decidable_rel
+-/
+attribute [local instance] LinearOrderedAddCommGroup.decidable_lt
+
+theorem sign_sum {ι : Type _} {s : Finset ι} {f : ι → α} (hs : s.Nonempty) (t : SignType)
+    (h : ∀ i ∈ s, sign (f i) = t) : sign (∑ i in s, f i) = t := by
+  cases t
+  · simp_rw [zero_eq_zero, sign_eq_zero_iff] at h⊢
+    exact Finset.sum_eq_zero h
+  · simp_rw [neg_eq_neg_one, sign_eq_neg_one_iff] at h⊢
+    exact Finset.sum_neg h hs
+  · simp_rw [pos_eq_one, sign_eq_one_iff] at h⊢
+    exact Finset.sum_pos h hs
+#align sign_sum sign_sum
+
+end LinearOrderedAddCommGroup
+
+namespace Int
+
+theorem sign_eq_sign (n : ℤ) : Int.sign n = SignType.sign n := by
+  obtain (n | _) | _ := n <;> simp [sign, Int.sign_neg, negSucc_lt_zero]
+#align int.sign_eq_sign Int.sign_eq_sign
+
+end Int
+
+open Finset Nat
+
+open BigOperators
+
+/- Porting note: For all the following theorems, needed to add {α : Type u_1} to the assumptions
```

Comment:
why can't they live in two different universes?

Reply:
if we have to change the proof definitely we should just leave it be, but I do wonder what's the "best" statement of this theorem. `beta` can live in `Type`, if we want, as it is a `fintype` and therefore small. cc @YaelDillies who I think wrote this

Reply:
This is used in LTE and there everything must be in the same universe (because categories). If you want to generalize the statement, then you would need to first get `β` down to `Type` then back up to `Type v`.